common.rkt

```diff
@@ -23,50 +25,181 @@
+  (let* ((sub (state-sub st)) (tm (walk tm sub)))
     (if (pair? tm)
-        `(,(walk* (car tm) sub) .  ,(walk* (cdr tm) sub))
+        `(,(walk* (car tm) st) .  ,(walk* (cdr tm) st))
         tm)))
 (define (reified-index index)
   (string->symbol
     (string-append "_." (number->string index))))
 (define (reify tm st)
   (define index -1)
-  (walk* tm (let loop ((tm tm) (sub (state-sub st)))
-              (define t (walk tm sub))
-              (cond ((pair? t) (loop (cdr t) (loop (car t) sub)))
-                    ((var? t)  (set! index (+ 1 index))
-                               (extend-sub t (reified-index index) sub))
-                    (else      sub)))))
+  (let ((results (let loop ((tm tm) (st st))
+                   (define t (walk tm (state-sub st)))
+                   (cond ((pair? t) (loop (cdr t) (loop (car t) st)))
+                         ((var? t)  (set! index (+ 1 index))
+                                    (state (extend-sub t (reified-index index) (state-sub st)) (state-diseq st) (state-types st)))
+                         (else      st)))))
+    (let* ((walked-sub (walk* tm results))
+           (diseq (map (lambda (=/=) (cons (length =/=) =/=)) (state-diseq st)))
+           (diseq (map cdr (sort diseq (lambda (x y) (< (car x) (car y))))))
+           (st (diseq-simplify (state (state-sub st) diseq (state-types st))))
+           (diseq (walk* (map pretty-diseq (state-diseq st)) results))
+           (diseq (filter-not contains-fresh? diseq))
+           (diseq (map (lambda (=/=) (sort =/= term<?)) diseq))
+           (diseq (if (null? diseq) '() (list (cons '=/= diseq))))
+           (types (walk* (map pretty-types (state-types st)) results))
+           (types (filter-not contains-fresh? types))
+           (cxs (append types diseq)))
+      (if (null? cxs)
+          walked-sub
+          (Ans walked-sub (sort cxs term<?))))))
+
 (define (reify/initial-var st)
   (reify initial-var st))
+
+(define (term<? u v)
+  (eqv? (term-compare u v) -1))
+
+(define (term-compare u v)
+  (cond
+    ((eqv? u v) 0)
+    ((null? u) -1)
+    ((null? v) 1)
+    ((not u) -1)
+    ((not v) 1)
+    ((eqv? u #t) -1)
+    ((eqv? v #t) 1)
+    ((number? u) (if (number? v) (if (< u v) -1 1) -1))
+    ((number? v) 1)
+    ((symbol? u) (if (symbol? v) (if (symbol<? u v) -1 1) -1))
+    ((symbol? v) 1)
+    ((string? u) (if (string? v) (if (string<? u v) -1 (if (string=? u v) 0 1)) -1))
+    ((string? v) 1)
+    ((pair? u) (if (pair? v) 
+                   (let ((compared-cars (term-compare (car u) (car v))))
+                     (if (eqv? compared-cars 0)
+                         (term-compare (cdr u) (cdr v))
+                         compared-cars))
+                   -1))
+    ((pair? v) 1)
+    (else 1)))
+
+(define (contains-fresh? x)
+  (if (pair? x)
+      (or (contains-fresh? (car x)) (contains-fresh? (cdr x)))
+      (var? x)))
+(define (pretty-diseq =/=s) (map (lambda (=/=) (list (car =/=) (cdr =/=))) =/=s))
```

Comment:
Since `=/=` is symmetric, let's sort its components so that the smaller one is always on the left.

Reply:
If we do this then if we have (=/= x 'm), then the result will be `((=/= ((m _.0)))` instead of `((=/= ((_.0 m)))`, are we okay with that?

Reply:
wait ignore that, I think I have something backwards

Reply:
I changed the sort order so that symbols come before numbers. that way we will have stuff like  `(_.0 12)` instead of `(12 _.0)`